tfx/experimental/pipeline_testing/dummy_component_launcher.py

```diff
@@ -0,0 +1,102 @@
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import abc
+from typing import Any, Dict, List, Text, Type
+import absl
+
+from tfx import types
+from tfx.components.base import base_executor
+from tfx.orchestration.launcher.in_process_component_launcher import InProcessComponentLauncher
+from tfx.experimental.pipeline_testing.dummy_executor import BaseDummyExecutor
+
+class BaseDummyComponentLauncher(InProcessComponentLauncher):
+  """Responsible for launching a dummy executor.
+
+  The executor will be launched in the same process of the rest of the
+  component, i.e. its driver and publisher.
+  """
+  def __init__(self, **kwargs):
+    super(BaseDummyComponentLauncher, self).__init__(**kwargs)
+    self._record_dir = os.path.join(os.environ['HOME'],
+                                    'tfx/tfx/experimental/pipeline_testing/',
+                                    'testdata')
+    self._executors_dict = {}
+    self._dummy_component_id = []
+    if not os.path.exists(self._record_dir):
+      raise Exception("Must record input/output in {}".format(self._record_dir))
+
+  @abc.abstractmethod
+  def set_dummy_executors(self, component_ids: List[Text],
+                          custom_executor_dict: base_executor.BaseExecutor):
+    """
+    component_ids: list of component_id to set a dummy executor
+    custom_executor_dict: dictionary holding user custom executors
+    """
+    pass
+
+  def _run_executor(self, execution_id: int,
+                    input_dict: Dict[Text, List[types.Artifact]],
+                    output_dict: Dict[Text, List[types.Artifact]],
+                    exec_properties: Dict[Text, Any]) -> None:
+    """Execute underlying component implementation."""
+    component_id = self._component_info.component_id
+
+    executor_context = base_executor.BaseExecutor.Context(
+        beam_pipeline_args=self._beam_pipeline_args,
+        tmp_dir=os.path.join(self._pipeline_info.pipeline_root, '.temp', ''),
+        unique_id=str(execution_id))
+    if component_id in self._dummy_component_id:
+      executor = BaseDummyExecutor(component_id,
+                                   self._record_dir,
+                                   executor_context)
+      executor.Do(input_dict, output_dict, exec_properties)
+    elif component_id in self._executors_dict.keys():
+      executor = self._executors_dict[component_id](executor_context)
+      executor.Do(input_dict, output_dict, exec_properties)
+    else:
+      super(BaseDummyComponentLauncher, self)._run_executor(execution_id,
+                                                            input_dict,
+                                                            output_dict,
+                                                            exec_properties)
+
+
+class MyDummyComponentLauncher(BaseDummyComponentLauncher):
```

Comment:
My...Launcher has to be placed in the examples, because it's for chicago taxi pipeline examples.

Reply:
I encountered a pickling error when I moved MyDummyComponentLauncher to taxi_pipeline_regression_test.py (https://gist.github.com/iamsusiep/f4cd1b9c1955ae6a61058c37c58522b4)

Reply:
If we remove `__init__` function of `My...Launcher`, it might work. Can you check this?